platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/Bucket.java

```diff
@@ -246,242 +216,233 @@ public long findValue(final int keyHashCode, final K key, final long notFoundVal
-            for (int i = 0; i < entryCount; i++) {
-                final int keySize = getKeySize(entryOffset);
-                bucketBuffer.position(entryOffset);
-                final int readHash = bucketBuffer.getInt();
-                final long value = bucketBuffer.getLong();
-                final K key = getKey(entryOffset);
-                sb.append("    ENTRY["
-                        + i
-                        + "] value= "
-                        + value
-                        + " keyHashCode="
-                        + readHash
-                        + " keyVer="
-                        + keySerializationVersion
-                        + " key="
-                        + key
-                        + " keySize="
-                        + keySize
-                        + "\n");
-                entryOffset += KEY_HASHCODE_SIZE + VALUE_SIZE + keySize;
-            }
-        } catch (IOException e) {
-            logger.error(EXCEPTION.getMarker(), "Failed enumerating bucket entries", e);
-        }
-        bucketBuffer.clear();
-        sb.append("} RAW DATA = ");
-        final byte[] bucketArray = bucketBuffer.array();
-        for (int i = 0; i < size; i++) {
-            sb.append(String.format("%02X ", bucketArray[i]).toUpperCase());
+        for (int i = 0; i < entryCount; i++) {
+            final BucketEntry entry = entries.get(i);
+            final int hashCode = entry.getHashCode();
+            final long value = entry.getValue();
+            final K key = entry.getKey();
+            sb.append("    ENTRY[" + i + "] value= " + value + " keyHashCode=" + hashCode + " key=" + key + "\n");
         }
+        sb.append("}");
         return sb.toString();
     }
 
-    // =================================================================================================================
-    // FindResult class
-
     /**
-     * Simple record for entry lookup results. If an entry is found, "found" is set to true,
-     * "entryOffset" is the entry offset in bytes in the bucket buffer, "entryIndex" is the
-     * entry index in the array of entries, and "entryValue" is the entry value. If no entity
-     * is found, "found" is false, "entryOffset" is the total size of the bucket buffer,
-     * "entryIndex" and "entryValue" are undefined.
+     * A single entry in a bucket, which contains key hash code, value (usually, path), and
+     * full serialized key. A bucket may contain multiple such entries.
+     *
+     * <p>This class would be a record, if it was immutable. However, when a value is updated
+     * in a bucket, and a bucket entry already exists for the same key, instead of creating
+     * a new entry, we just update the value in the existing entry.
      */
-    private record FindResult(int entryOffset, int entryIndex, boolean found, long entryValue) {}
+    private class BucketEntry {
+
+        /** Key hash code */
+        private int hashCode;
```

Comment:
It looks like these fields should be final

Reply:
I agree, both `hashCode` and `key` can be final, while `value` may be updated on Bucket.put()